data_rentgen/server/services/lineage.py

```diff
@@ -1115,12 +1127,181 @@ async def _dataset_lineage_with_job_granularity(
 
         return result
 
+    async def _dataset_lineage_with_dataset_granularity(  # noqa: C901
+        self,
+        datasets_by_id: dict[int, Dataset],
+        dataset_symlinks_by_id: dict[tuple[int, int], DatasetSymlink],
+        direction: LineageDirectionV1,
+        since: datetime,
+        until: datetime | None,
+        depth: int,
+        level: int,
+        ids_to_skip: IdsToSkip | None = None,
+    ) -> LineageServiceResult:
+        result = LineageServiceResult(
+            datasets=datasets_by_id,
+            dataset_symlinks=dataset_symlinks_by_id,
+        )
+        ids_to_skip = ids_to_skip or IdsToSkip()
+        ids_to_skip = ids_to_skip.merge(IdsToSkip.from_result(result))
+        layer_datasets_by_id = {
+            "downstream": datasets_by_id,
+            "upstream": datasets_by_id,
```

Comment:
Both keys lead to the same object, so adding new items to any of them will actually add item to both